File: scripts/run_convolution_tests.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <math.h>

// Standalone convolution test runner (links to convolution.c)

extern void convolution_test(int M, int N, const char *filter_type);

int main() {
    printf("=== Convolution Performance Tests ===\n\n");
    
    // Test different image sizes (M) and filter sizes (N)
    int image_sizes[] = {256, 512, 1024};
    int filter_sizes[] = {3, 5, 7};
    const char *filters[] = {"sobel_x", "gaussian", "sharpen"};
    
    printf("| Image Size (M) | Filter Size (N) | Filter Type | Time (seconds) |\n");
    printf("|----------------|-----------------|-------------|----------------|\n");
    
    for (int m_idx = 0; m_idx < 3; m_idx++) {
        int M = image_sizes[m_idx];
        for (int n_idx = 0; n_idx < 3; n_idx++) {
            int N = filter_sizes[n_idx];
            for (int f_idx = 0; f_idx < 3; f_idx++) {
                const char *filter = filters[f_idx];
                
                // Build command
                char cmd[256];
                snprintf(cmd, sizeof(cmd), "convolution.exe %d %d %s", M, N, filter);
                
                printf("Running: %s\n", cmd);
                system(cmd);
                printf("\n");
            }
        }
    }
    
    return 0;
}

File: scripts/run_convolution_tests.py
"""Run CPU and GPU convolution programs for several image sizes and filter types."""
import os
import subprocess
import sys

script_dir = os.path.dirname(os.path.abspath(__file__))
root_dir = os.path.dirname(script_dir)
is_windows = sys.platform.startswith("win")
ext = ".exe" if is_windows else ""

image_sizes = [256, 512, 1024]
filter_sizes = [3]
filter_types = ["sobel_x", "sobel_y", "gaussian", "sharpen", "laplacian"]

def exe(name):
    return name + ext

def run(cmd):
    exe_path = os.path.join(root_dir, cmd[0])
    if not os.path.isfile(exe_path):
        print(f"(Skipping: {cmd[0]} not found. Build it first.)")
        return
    full_cmd = [exe_path] + list(cmd[1:])
    subprocess.run(full_cmd, cwd=root_dir)

print("=== CONVOLUTION ===\n")

print("--- CPU Convolution ---")
for m in image_sizes:
    for n in filter_sizes:
        for f in filter_types:
            run([exe("convolution"), str(m), str(n), f])

print("\n--- GPU Convolution ---")
for m in image_sizes:
    for n in filter_sizes:
        for f in filter_types:
            run([exe("convolution_gpu"), str(m), str(n), f])

print("\nConvolution tests done.")
